### DISCUSSION TITLE
Threading and rendering architecture

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#80] [Ideas]
Suggestions for how best to do threading?

**Comment by awolven** [comment]
I can comment on task-oriented threading, data oriented threading should be
self explanatory.

The main thread, and because of Cocoa restrictions, should run the render
loop.

The render loop should be dumb.  You should not be able to run arbitrary
code in the render thread.
The render thread should just render draw lists, essentially.

You can have a second thread that has an input queue for commands that are
more complicated than can be done with a callback.
Since glfwPollEvents must happen in the render thread, all callbacks should
not ever enter the debugger.  They can print a backtrace if necessary and
let the render loop continue.
Complex commands should be handled with a callback that places events in
the input queue for the second thread to pick up.  All code that can be
compiled or executed at runtime should happen in the second thread or any
other thread but not in the render thread.  Commands that change the
structure of a graphical object should enqueue the draw list from the
graphical object to an update queue on the render thread, which are picked
up and processed before the draw list is rendered.

This architecture prevents random code from crashing the app.  A secondary
or tertiary thread can enter the debugger without stopping rendering.  This
means, with a graphical debugger the error can be handled in the graphics
ui potentially instead of just in emacs, if you so desire.

**Reply by kaveh808** [reply]
Sounds good.

**Reply by foretspaisibles** [reply]
So for instance a part of the scene that has still not be computed (in an interactive editing session) could be rendered as a wireframe or a box until other computation have been done and more details are available to the rendering loop?

**Reply by awolven** [reply]
if you want to do it that way, sure.

**Comment by kaveh808** [comment]
Updated the topic so we can discuss architecture related matters here.

**Reply by awolven** [reply]
this thread should be merged with the concurrency thread

**Reply by kaveh808** [reply]
I think this topic is important enough to have its own thread.

**Comment by awolven** [comment]
I guess I intended my post about the render thread and the input thread and
tertiary threads to be in the more "concurrency" architecture discussions.

**Reply by kaveh808** [reply]
No worries. I don't know how to merge discussions. Why don't you post your comment above in the concurrency discussion and I'll close this one.
